Mathlib/CategoryTheory/MorphismProperty/Tactic.lean

```diff
@@ -0,0 +1,231 @@
+-/
+def addMorphismPropertyInstancesAux (classTerm : TSyntax `term) (verbose : Bool := false) :
+    Command.CommandElabM PUnit := Command.runTermElabM fun fvars ↦ do
+  let env ← getEnv
+  -- Get the set of lemmas tagged with `thmAttr`
+  let lemmas := (thmAttr.ext.getState env).fold (·.push (·, ·))
+    (thmAttr.ext.toEnvExtension.getState env).importedEntries.flatten
+  let mut logMsg : MessageData := ""
+  let mut success := 0
+  -- `classExpr` is the morphism property class, elaborated as an `Expr`.
+  let classExpr ← Term.elabTerm classTerm none
+  -- We first check if the type of `classExpr` is defeq to `MorphismProperty _`.
+  do
+    let mp ← mkConstWithFreshMVarLevels ``MorphismProperty
+    let (args, _, _) ← forallMetaTelescope (← inferType mp)
+    let mp ← mkAppOptM ``MorphismProperty (args.map some)
+    unless ← isDefEq mp (← inferType classExpr) do
+      throwError m!"{classExpr} is not of type `MorphismProperty _`."
+  -- We only continue if `classExpr` is the application of some
+  -- `morphismPropertyClass` to some arguments.
+  if let .const morphismPropertyClass _lvl := classExpr.getAppFn then
+    for (lemmaName, config) in lemmas do
+      try
+        -- First we check if the lemma already exists.
+        let name := .str morphismPropertyClass lemmaName.toString
+        checkNotAlreadyDeclared name
+        -- We attempt to feed `classExpr` into `lemmaName` and fill in all typeclass arguments
+        -- of `lemmaName` about `classExpr`.
+        let classExpr ← Term.elabTerm classTerm none
+        let lemmaExpr ← mkConstWithFreshMVarLevels lemmaName
+        let args ← getArgs lemmaExpr classExpr
+        let proofTerm ← mkAppOptM' lemmaExpr (args.map some)
+        -- We run through the type of the constructed `proofTerm` and convert
+        -- all type class arguments into `instImplicit`.
+        let statement ← forallTelescope (← inferType proofTerm) fun xs ty => do
+          let xs' := xs.filterMapM fun x ↦ do
+            if (← isClass? (← inferType x)).isSome then
+              return (x.fvarId!, BinderInfo.instImplicit)
+            else
+              return none
+          withNewBinderInfos (← xs') <| liftMetaM do
+            mkForallFVars xs ty
+        -- We add back all the scoped variable used in the statement and proof.
+        let fvars := (← Term.collectUnassignedMVars proofTerm) ++ fvars
+        let proofTerm ← instantiateMVars (← Term.levelMVarToParam
+          (← mkLambdaFVars (usedOnly := true) fvars proofTerm))
+        let statement ← instantiateMVars (← mkForallFVars (usedOnly := true) fvars statement)
+        let decl : TheoremVal :=
+        { name := name
+          levelParams := (collectLevelParams {} proofTerm).params.toList
+          type := statement
+          value := proofTerm }
+        -- We add the declaration and add it as an instance
+        addAndCompile (.thmDecl decl)
```

Comment:
Is there any change the theorem being added could attempt to compile and fail? It seems odd to try to compile it here if it is a theorem, do you have an application of that in mind?